### PR TITLE
Support record load progress for resume loading

### PR DESCRIPTION
**Pull request by Linary**
Implement #69

Change-Id: I39578958f0915f42840bf2789369df647b4867a0

### CODE REVIEW

File: src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java [hunk 1]
```diff
@@ -151,40 +171,55 @@ private void createSchema() {
...
         return summary;
     }
 
+    private void loadVertices() {
+        // Update progress
+        progress.loadingElemType(ElemType.VERTEX);
+        // Execute loading tasks
+        List<VertexSource> vertexSources = this.graphSource.vertexSources();
+        for (VertexSource source : vertexSources) {
+            String uniqueKey = source.uniqueKey();
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java]: label is not unique

File: src/main/java/com/baidu/hugegraph/loader/reader/file/AbstractFileReader.java [hunk 2]
```diff
@@ -67,10 +70,17 @@ public FileSource source() {
...
+        return this.progress;
+    }
+
     protected abstract Readers openReaders() throws IOException;
 
     @Override
-    public void init() {
+    public void init(LoadProgress progress) {
+        assert progress.inputSource() == null;
+        this.progress = new ReadableProgress();
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/file/AbstractFileReader.java]: FileProgress

File: src/main/java/com/baidu/hugegraph/loader/reader/file/AbstractFileReader.java [hunk 3]
```diff
@@ -67,10 +70,17 @@ public FileSource source() {
...
+        progress.inputSource(this.progress);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/file/AbstractFileReader.java]: only set when finished

File: src/main/java/com/baidu/hugegraph/loader/progress/ElementProgress.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * under the License.
+ */
+
+package com.baidu.hugegraph.loader.progress;
+
+import java.util.LinkedHashMap;
+
+import com.baidu.hugegraph.loader.source.graph.ElementSource;
+
+public final class ElementProgress extends LinkedHashMap<String, InputProgress> {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/progress/ElementProgress.java]: InputProgressMap

File: src/main/java/com/baidu/hugegraph/loader/reader/file/FileItem.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class FileItem extends InputItem {
+
+    @JsonProperty("name")
+    private final String name;
+    @JsonProperty("last_modified")
+    private final long timestamp;
+    @JsonProperty("check_sum")
+    private final long checkSum;
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/file/FileItem.java]: checksum

File: src/main/java/com/baidu/hugegraph/loader/reader/file/FileItem.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class FileItem extends InputItem {
+
+    @JsonProperty("name")
+    private final String name;
+    @JsonProperty("last_modified")
+    private final long timestamp;
+    @JsonProperty("check_sum")
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/file/FileItem.java]: checksum

File: src/main/java/com/baidu/hugegraph/loader/summary/LoadCounter.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+package com.baidu.hugegraph.loader.summary;
+
+import java.util.concurrent.atomic.LongAdder;
+
+public final class LoadCounter {
+
+    // Modified in only one thread
+    private long parseFailureNum;
+    private final LongAdder insertSuccessNum;
```
[the remark this answers is not in the document]
Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/summary/LoadCounter.java]: LongAdder has better perf than AtomicLong

File: assembly/static/conf/log4j2.xml [hunk 8]
```diff
@@ -55,14 +55,14 @@
...
         </logger>
-        <logger name="vertexInsertError" level="INFO" additivity="false">
-            <appender-ref ref="vertexInsertErrorFile"/>
+        <logger name="vertex-insert-error" level="INFO" additivity="false">
+            <appender-ref ref="vertex-insert-error-file"/>
         </logger>
-        <logger name="edgeInsertError" level="INFO" additivity="false">
-            <appender-ref ref="edgeInsertErrorFile"/>
+        <logger name="edge-insert-error" level="INFO" additivity="false">
+            <appender-ref ref="edge-insert-error-file"/>
```
Comment (javeme) [inline assembly/static/conf/log4j2.xml]: no empty line at the end of file

File: src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java [hunk 9]
```diff
@@ -77,162 +66,93 @@ public static void main(String[] args) {
...
-        // Shutdown task manager
-        this.taskManager.shutdown(this.options.shutdownTimeout);
-        LoadUtil.exit(code);
-    }
-
-    private void createSchema() {
-        File schemaFile = FileUtils.getFile(this.options.schema);
-        HugeClient client = HugeClientWrapper.get(this.options);
+    private void createSchema(LoadContext context) {
+        LoadOptions options = context.options();
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java]: use this.context

File: src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java [hunk 10]
```diff
@@ -77,162 +66,93 @@ public static void main(String[] args) {
...
-        Printer.printElemType(type);
-        Instant beginTime = Instant.now();
-
-        if (type.isVertex()) {
-            this.loadVertices();
-        } else {
-            assert type.isEdge();
-            this.loadEdges();
-        }
+    private void loadVertices(LoadContext context) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java]: ditto

File: src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java [hunk 11]
```diff
@@ -77,162 +66,93 @@ public static void main(String[] args) {
...
-    private void stopLoading(int code) {
-        // Shutdown task manager
-        this.taskManager.shutdown(this.options.shutdownTimeout);
-        LoadUtil.exit(code);
-    }
-
-    private void createSchema() {
-        File schemaFile = FileUtils.getFile(this.options.schema);
-        HugeClient client = HugeClientWrapper.get(this.options);
+    private void createSchema(LoadContext context) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java]: unnecessary to pass context

File: src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java [hunk 12]
```diff
@@ -247,44 +167,50 @@ private void loadVertex(VertexBuilder builder) {
         }
     }
 
-    private void loadEdges() {
-        // Used to get and update progress
-        ElementProgress oldEdgeProgress = this.oldProgress.edge();
-        ElementProgress newEdgeProgress = this.newProgress.edge();
+    private void loadEdges(LoadContext context) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java]: ditto

File: src/main/java/com/baidu/hugegraph/loader/LoadContext.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+import com.baidu.hugegraph.loader.util.LoadUtil;
+import com.baidu.hugegraph.util.E;
+import com.baidu.hugegraph.util.Log;
+import com.beust.jcommander.JCommander;
+
+public final class LoadContext {
+
+    private static final Logger LOG = Log.logger(HugeGraphLoader.class);
+
+    private final LoadOptions loadOptions;
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/LoadContext.java]: loadOptions => options

File: src/main/java/com/baidu/hugegraph/loader/executor/LoadOptions.java [hunk 14]
```diff
@@ -57,7 +57,7 @@
 
     @Parameter(names = {"--incremental-mode"}, arity = 1,
               description = "Load data from the breakpoint of last time")
-    public boolean incrementalMode = true;
+    public boolean incrementalMode = false;
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/executor/LoadOptions.java]: keep incrementalMode=true

File: src/main/resources/log4j2.xml [hunk 15]
```diff
@@ -40,29 +40,30 @@
...
         </logger>
-        <logger name="vertexInsertError" level="INFO" additivity="false">
-            <appender-ref ref="vertexInsertErrorFile"/>
+        <logger name="vertex-insert-error" level="INFO" additivity="false">
+            <appender-ref ref="vertex-insert-error-file"/>
         </logger>
-        <logger name="edgeInsertError" level="INFO" additivity="false">
-            <appender-ref ref="edgeInsertErrorFile"/>
+        <logger name="edge-insert-error" level="INFO" additivity="false">
+            <appender-ref ref="edge-insert-error-file"/>
```
Comment (javeme) [inline src/main/resources/log4j2.xml]: add empty line

File: src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java [hunk 16]
```diff
@@ -85,14 +86,15 @@ public InputReader reader() {
...
+    private void progress(ElementSource source, LoadContext context) {
         LoadProgress oldProgress = context.oldProgress();
         LoadProgress newProgress = context.newProgress();
-        ElementSource source = this.source();
         InputProgress oldInputProgress = oldProgress.get(this.type(), source);
         if (oldInputProgress == null) {
             oldInputProgress = new InputProgress(source);
         }
+        // Update loading element source
+        newProgress.addSource(this.type(), source);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java]: prefer to do it before constructing ElementBuilder

File: src/main/java/com/baidu/hugegraph/loader/progress/InputProgressMap.java [hunk 17]
```diff
@@ -22,14 +22,17 @@
 import java.util.HashMap;
 
 import com.baidu.hugegraph.loader.source.graph.ElementSource;
+import com.baidu.hugegraph.util.E;
 
 public final class InputProgressMap extends HashMap<String, InputProgress> {
 
     public InputProgress get(ElementSource source) {
+        E.checkNotNull(source, "element source");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/progress/InputProgressMap.java]: getBySource

File: src/main/java/com/baidu/hugegraph/loader/executor/LoadOptions.java (hunk 14 above)
Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/executor/LoadOptions.java]: I will modify it later

File: src/main/java/com/baidu/hugegraph/loader/reader/hdfs/HdfsFileReader.java [hunk 18]
```diff
@@ -92,7 +91,7 @@ protected Readers openReaders() throws IOException {
                 paths.add(new HdfsFile(this.hdfs, subPath));
             }
         }
-        return new Readers(this.source(), paths);
+        this.readers().readables(paths);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/hdfs/HdfsFileReader.java]: I think the original way(direct construction with List<Readable>) is easier to understand

File: src/main/java/com/baidu/hugegraph/loader/reader/file/Readers.java [hunk 19]
```diff
@@ -43,7 +42,7 @@
 import com.baidu.hugegraph.util.Log;
 
 /**
- * Used to iterate all readable data source, like local files, hdfs paths
+ * Used to iterate all readable data desc, like local files, hdfs paths
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/file/Readers.java]: iterate all readable data files

File: src/main/java/com/baidu/hugegraph/loader/source/desc/EdgeDesc.java [hunk 20]
```diff
@@ -17,34 +17,40 @@
...
+import com.baidu.hugegraph.loader.constant.ElemType;
 import com.baidu.hugegraph.util.E;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
-public class EdgeSource extends ElementSource {
+public class EdgeDesc extends ElementDesc {
 
-    @JsonProperty("source")
+    @JsonProperty("desc")
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/source/desc/EdgeDesc.java]: keep source

File: src/main/java/com/baidu/hugegraph/loader/source/desc/EdgeDesc.java [hunk 21]
```diff
@@ -17,34 +17,40 @@
...
     private final List<String> sourceFields;
     @JsonProperty("target")
     private final List<String> targetFields;
 
     @JsonCreator
-    public EdgeSource(@JsonProperty("source") List<String> sourceFields,
-                      @JsonProperty("target") List<String> targetFields) {
+    public EdgeDesc(@JsonProperty("desc") List<String> sourceFields,
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/source/desc/EdgeDesc.java]: ditto

File: src/main/java/com/baidu/hugegraph/loader/source/desc/EdgeDesc.java [hunk 22]
```diff
@@ -17,34 +17,40 @@
...
+        return ElemType.EDGE;
+    }
+
     @Override
     public void check() throws IllegalArgumentException {
         super.check();
         E.checkArgument(this.sourceFields != null &&
                         !this.sourceFields.isEmpty(),
-                        "The source field of edge label '%s' " +
+                        "The desc field of edge label '%s' " +
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/source/desc/EdgeDesc.java]: ditto

File: src/main/java/com/baidu/hugegraph/loader/source/desc/GraphDesc.java [hunk 23]
```diff
@@ -39,56 +39,56 @@
...
         File file = FileUtils.getFile(options.file);
         try {
             String json = FileUtils.readFileToString(file, Constants.CHARSET);
-            GraphSource source = JsonUtil.fromJson(json, GraphSource.class);
+            GraphDesc source = JsonUtil.fromJson(json, GraphDesc.class);
             source.check();
             return source;
         } catch (IOException | IllegalArgumentException e) {
-            throw new LoadException("Read graph source file '%s' error",
+            throw new LoadException("Read graph desc file '%s' error",
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/source/desc/GraphDesc.java]: description

File: src/main/java/com/baidu/hugegraph/loader/source/desc/GraphDesc.java [hunk 24]
```diff
@@ -39,56 +39,56 @@
...
 
     @Override
     public void check() throws IllegalArgumentException {
-        LOG.info("Checking vertex sources");
-        this.vertexSources.forEach(VertexSource::check);
-        this.checkNoSameSource(this.vertexSources);
-        LOG.info("Checking edge sources");
-        this.edgeSources.forEach(EdgeSource::check);
-        this.checkNoSameSource(this.edgeSources);
+        LOG.info("Checking vertex descs");
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/source/desc/GraphDesc.java]: ditto

File: src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java [hunk 25]
```diff
@@ -171,13 +178,13 @@ private void loadEdges() {
...
-        List<EdgeSource> edgeSources = this.graphSource.edgeSources();
-        for (EdgeSource source : edgeSources) {
-            // Update loading edge source
-            newProgress.addSource(source);
-            LOG.info("Loading edge source '{}'", source.label());
+        List<EdgeDesc> edgeDescs = this.graphDesc.edgeDescs();
+        for (EdgeDesc desc : edgeDescs) {
+            // Update loading edge desc
+            newProgress.addDesc(desc);
+            LOG.info("Loading edge desc '{}'", desc.label());
```
[the remark this answers is not in the document]
Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/HugeGraphLoader.java]: InputSource not certainly has path, such as JDBCSource

File: src/main/java/com/baidu/hugegraph/loader/LoadContext.java [hunk 26]
```diff
@@ -81,13 +81,24 @@ private static LoadOptions parseCheckOptions(String[] args) {
...
+                        "Must specified struct description file");
+        File structFile = new File(options.file);
+        if (!structFile.canRead()) {
+            LOG.error("Struct file must be readable: '{}'",
+                      structFile.getAbsolutePath());
             LoadUtil.exitWithUsage(commander, Constants.EXIT_CODE_ERROR);
         }
+
+        // Check option "-s"
+        E.checkArgument(!StringUtils.isEmpty(options.schema),
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/LoadContext.java]: not required and add a test without schema file

File: src/main/java/com/baidu/hugegraph/loader/reader/file/FileReader.java [hunk 27]
```diff
@@ -81,20 +81,20 @@ public void init(LoadContext context, ElementDesc desc) {
                 this.parser.parseHeader(headerLine);
             } else {
                 throw new LoadException("Failed to read header from " +
-                                        "file desc '%s'", this.source);
+                                        "file struct '%s'", this.source);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/file/FileReader.java]: file source

File: src/main/java/com/baidu/hugegraph/loader/task/BatchInsertTask.java [hunk 28]
```diff
@@ -22,51 +22,39 @@
...
     public Integer get() {
-        if (this.batch == null || this.batch.isEmpty()) {
-            return 0;
-        }
-
         int retryCount = 0;
         do {
             try {
-                this.execute();
+                HugeClientWrapper.addBatch(this.type, this.batch,
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/BatchInsertTask.java]: use this.client, maybe HugeClientWrapper has not been initialized yet

File: src/main/java/com/baidu/hugegraph/loader/task/BatchInsertTask.java [hunk 29]
```diff
@@ -46,6 +46,8 @@ public BatchInsertTask(ElemType type, List<GE> batch, LoadOptions options) {
         this.type = type;
         this.batch = batch;
         this.options = options;
+        // Ensure internal client initialized
+        HugeClientWrapper.get(options);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/BatchInsertTask.java]: prefer to use this.client for readability

File: src/main/java/com/baidu/hugegraph/loader/task/SingleInsertTask.java [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+                LOG.error("Single insert {} error", type, e);
+                if (options.testMode) {
+                    throw e;
+                }
+                this.failureLogger.error(type, new InsertException(element, e));
+
+                if (metrics.insertFailure() >= options.maxInsertErrors) {
+                    Printer.printError("More than %s %s insert error... " +
+                                       "stopping",
+                                       options.maxInsertErrors, type);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/SingleInsertTask.java]: move to previous line

File: src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java [hunk 31]
```diff
@@ -166,35 +147,9 @@ public void shutdown() {
...
-                    }
-                }
-            }
-            Printer.printProgress(metrics, SINGLE_PRINT_FREQ, batch.size());
-        };
-
+        InsertTask<GE> task = new SingleInsertTask<>(this.context, type, batch);
         CompletableFuture<Void> future;
-        future = CompletableFuture.runAsync(singleInsertTask, this.singleService);
+        future = CompletableFuture.runAsync(task, this.singleService);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]:
```
CompletableFuture.runAsync(task, this.singleService)
                 .whenComplete(...)
```

File: src/main/java/com/baidu/hugegraph/loader/util/HugeClientWrapper.java [hunk 32]
```diff
@@ -61,28 +54,10 @@ private static HugeClient newHugeClient(LoadOptions options) {
...
-    @SuppressWarnings("unchecked")
-    public static <GE extends GraphElement> GE addSingle(ElemType type,
-                                                         GE element) {
-        GraphManager graph = instance.graph();
-        if (type.isVertex()) {
-            return (GE) graph.addVertex((Vertex) element);
-        } else {
-            assert type.isEdge();
-            return (GE) graph.addEdge((Edge) element);
+    public static void close() {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/HugeClientWrapper.java]: add lock:
synchronized(HugeClientWrapper.class)

File: src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java [hunk 33]
```diff
@@ -139,22 +128,14 @@ public void shutdown() {
                                     "batch in batch mode", e, type);
         }
 
-        BatchInsertTask<GE> task = new BatchInsertTask<>(type, batch,
-                                                         this.options);
-        CompletableFuture<Integer> future;
-        future = CompletableFuture.supplyAsync(task, this.batchService);
+        InsertTask<GE> task = new BatchInsertTask<>(this.context, type, batch);
+        CompletableFuture<Void> future;
+        future = CompletableFuture.runAsync(task, this.batchService);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]: move to previous line

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 34]
```diff
@@ -54,46 +52,96 @@ public static void printElemType(ElemType type) {
...
+     *
+     * struct 'user-e6d5731c'
+     * parse:
+     *     parse success           : 6040
+     *     parse failure           : 0
+     *     parse time(second)      : 0
+     * load:
+     *     load success            : 6040
+     *     load failure            : 0
+     *     load time(second)       : 0
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: missing load rate

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 35]
```diff
@@ -54,46 +52,96 @@ public static void printElemType(ElemType type) {
...
+     *     load failure            : 0
+     *     load time(second)       : 17
+     *     load rate               : 58835
+     * ------------------------------------------
+     * total metrics
+     * parse:
+     *     parse success           : 1010132
+     *     parse failure           : 0
+     *     parse time(second)      : 3
+     *     parse rate              : 336710
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: unnecessary to print parse success/time/rate, just print parse failure for total sumary

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 36]
```diff
@@ -54,46 +52,96 @@ public static void printElemType(ElemType type) {
...
+     *     parse failure           : 0
+     *     parse time(second)      : 3
+     *     parse rate              : 333403
+     * load:
+     *     load success            : 1000209
+     *     load failure            : 0
+     *     load time(second)       : 17
+     *     load rate               : 58835
+     * ------------------------------------------
+     * total metrics
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: prefer to divide into two parts: total vertices and  total edges

**Comment by Linary** [comment]
Updates and optimization points include
1. Save load progress in order to continue from the breakpoint at next loading(will take effect after the commit "failure logger ...")
2. Support for incremental loading
3. Use LoadContext to manage auxiliary objects
4. Optimize the metrics information
5. Combine vertex and edge processing logic
6. Split a separate single insertion task

File: src/main/java/com/baidu/hugegraph/loader/task/BatchInsertTask.java [hunk 37]
```diff
@@ -98,6 +98,8 @@ private int waitThenRetry(int retryCount, RuntimeException e) {
         }
 
         if (++retryCount > options.retryTimes) {
+            LOG.error("Batch insert has retried exceed {} times",
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/BatchInsertTask.java]: Batch insert has been retried more than {} times

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 38]
```diff
@@ -0,0 +1,216 @@
...
+     * total edge metrics
+     * parse:
+     *     parse success           : 1000209
+     *     parse failure           : 0
+     *     parse time(second)      : 3
+     *     parse rate              : 333403
+     * load:
+     *     load success            : 1000209
+     *     load failure            : 0
+     *     load time(second)       : 17
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: improve time unit

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 39]
```diff
@@ -0,0 +1,216 @@
...
+     * parse:
+     *     parse success           : 9923
+     *     parse failure           : 0
+     *     parse time(second)      : 0
+     *     parse rate              : -1
+     * load:
+     *     load success            : 9923
+     *     load failure            : 0
+     *     load time(second)       : 0
+     *     load rate               : -1
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: load rate(vertices/s)

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 40]
```diff
@@ -0,0 +1,216 @@
...
+     *     parse failure           : 0
+     *     load success            : 1000209
+     *     load failure            : 0
+     * ------------------------------------------
+     * total vertex metrics
+     * parse:
+     *     parse success           : 9923
+     *     parse failure           : 0
+     *     parse time(second)      : 0
+     *     parse rate              : -1
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: use milliseconds if seconds is < 1

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 41]
```diff
@@ -206,11 +188,24 @@ private static void printAndLog(String key, long value) {
...
+
     private static String backward(long count) {
         StringBuilder backward = new StringBuilder();
         for (int i = 0, len = String.valueOf(count).length(); i < len; i++) {
             backward.append("\b");
         }
         return backward.toString();
     }
+
+    public static String readableFormat(long time) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: rename to readableTime
and move to util and add some test cases

File: src/main/java/com/baidu/hugegraph/loader/reader/hdfs/HDFSFileReader.java [hunk 42]
```diff
@@ -126,39 +129,67 @@ private Configuration loadConfiguration() {
...
+        @Override
+        public InputItemProgress inputItemProgress() {
+            String name = this.path.getName();
+            long timestamp;
+            try {
+                timestamp = this.hdfs.getFileStatus(this.path)
+                                     .getModificationTime();
+            } catch (IOException e) {
+                throw new LoadException("Failed to get last modified time " +
+                                        "for hdfs path '%s'", e, this.path);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/reader/hdfs/HDFSFileReader.java]: replace all hdfs to HDFS

File: src/main/java/com/baidu/hugegraph/loader/reader/file/Readers.java [hunk 43]
```diff
@@ -0,0 +1,229 @@
...
+                header = this.reader.readLine();
+                start++;
+            }
+            for (long i = start; i < offset; i++) {
+                this.reader.readLine();
+            }
+        } catch (IOException e) {
+            throw new LoadException("Failed to skip the first %s lines " +
+                                    "of file %s, please ensure the file " +
+                                    "file must have at least %s lines");
```
Comment (zhoney) [inline src/main/java/com/baidu/hugegraph/loader/reader/file/Readers.java]: add args
delete one “file” in “file file”

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 44]
```diff
@@ -201,7 +201,8 @@ private static String backward(long count) {
         return backward.toString();
     }
 
-    public static String readableFormat(long time) {
+    // TODO: move to common util
+    public static String readableTime(long time) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: upgrade common version

File: src/main/java/com/baidu/hugegraph/loader/util/JsonUtil.java [hunk 45]
```diff
@@ -60,7 +70,47 @@ public static String toJson(Object object) {
         }
     }
 
+    public static <T> T fromJson(String json, TypeReference<?> typeRef) {
+        E.checkState(json != null, "Json value can't be null for '%s'",
+                     typeRef.getType());
+        try {
+            return MAPPER.readValue(json, typeRef);
+        } catch (IOException e) {
+            throw new SerializeException("Failed to deserialize json", e);
```
Comment (zhoney) [inline src/main/java/com/baidu/hugegraph/loader/util/JsonUtil.java]: DeserializeException

File: src/main/java/com/baidu/hugegraph/loader/util/JsonUtil.java [hunk 46]
```diff
@@ -60,7 +70,47 @@ public static String toJson(Object object) {
...
+    }
+
+    public static <K, V> Map<K, V> convertMap(String json, Class<K> kClazz,
+                                              Class<V> vClazz) {
+        JavaType type = MAPPER.getTypeFactory()
+                              .constructMapType(Map.class, kClazz, vClazz);
+        try {
+            return MAPPER.readValue(json, type);
+        } catch (IOException e) {
+            throw new SerializeException("Failed to deserialize json", e);
```
Comment (zhoney) [inline src/main/java/com/baidu/hugegraph/loader/util/JsonUtil.java]: ditto

File: src/main/java/com/baidu/hugegraph/loader/util/Printer.java [hunk 47]
```diff
@@ -0,0 +1,212 @@
...
+
+    private static String backward(long count) {
+        StringBuilder backward = new StringBuilder();
+        for (int i = 0, len = String.valueOf(count).length(); i < len; i++) {
+            backward.append("\b");
+        }
+        return backward.toString();
+    }
+
+    // TODO: move to common util
```
Comment (zhoney) [inline src/main/java/com/baidu/hugegraph/loader/util/Printer.java]: delete it and use hugegraph-common TimeUtil.readableTime()

File: src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java [hunk 48]
```diff
@@ -19,61 +19,71 @@
...
+import com.baidu.hugegraph.util.Log;
 
 public abstract class ElementBuilder<GE extends GraphElement>
                 implements AutoCloseableIterator<GE> {
 
-    private static final int VERTEX_ID_LIMIT = 128;
-    private static final String ID_CHARSET = "UTF-8";
+    private static final Logger LOG = Log.logger(ElementBuilder.class);
 
+    private final SchemaCache schema;
```
Comment (zhoney) [inline src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java]: why not singleton

File: src/main/java/com/baidu/hugegraph/loader/reader/file/Readers.java [hunk 49]
```diff
@@ -0,0 +1,229 @@
...
+            // The current file is read at the end, ready to read next one
+            this.close(true);
+            // Open the second or subsequent files
+            this.reader = this.openNext();
+            if (this.reader == null) {
+                return null;
+            }
+        }
+
+        this.newProgress.increaseLoadingOffset();
```
Comment (zhoney) [inline src/main/java/com/baidu/hugegraph/loader/reader/file/Readers.java]: adding offset before ensuring elements inserted may be not accurate when program exit(exceptions reaches limit)

File: src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java (hunk 48 above)
Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/builder/ElementBuilder.java]: singleton will lead tests share one cache, but some case should use different schema(same name)

File: src/main/java/com/baidu/hugegraph/loader/reader/file/Readers.java (hunk 49 above)
Comment (Linary) [reply src/main/java/com/baidu/hugegraph/loader/reader/file/Readers.java]: It's a good tip. I will be modify it in next commit.

File: src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java [hunk 50]
```diff
@@ -54,9 +54,8 @@
     public TaskManager(LoadContext context) {
         this.context = context;
         this.options = context.options();
-        this.batchSemaphore = new Semaphore(this.options.numThreads);
-        // Equivalent to have a buffer queue of length options.numThreads
-        this.singleSemaphore = new Semaphore(2 * this.options.numThreads);
+        this.batchSemaphore = new Semaphore(this.options.numThreads + 1);
+        this.singleSemaphore = new Semaphore(this.options.numThreads + 1);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/loader/task/TaskManager.java]: keep queue size 2x of this.options.numThreads

**Pull request merged**